[macOS + iOS Model] Compute bounding box when geometry is updated

mwyrzykowski opened the pull request:
#### 6cc828448d7b5f7c8b03f2752d79c2f5e41e3c3b
<pre>
[macOS + iOS Model] Compute bounding box when geometry is updated
<a href="https://bugs.webkit.org/show_bug.cgi?id=300410">https://bugs.webkit.org/show_bug.cgi?id=300410</a>
<a href="https://rdar.apple.com/162233357">rdar://162233357</a>

Reviewed by Etienne Segonzac.

Compute bounding box when the mesh geometry is updated.

Test: model-element/model-element-bounding-box.html passes locally after this change

* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
(WebCore::computeCenterAndExtents):
(WebCore::DDModelPlayer::load):

Canonical link: <a href="https://commits.webkit.org/302194@main">https://commits.webkit.org/302194@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b16f6c5f6ffe322f22db17e9753c4198b79f1727

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128115 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/399 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38947 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/135481 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79610 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129987 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/319 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/271 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97530 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/65423 "") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/01464882-a8cb-4bbd-b62d-e1ba5ab6bdd8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131063 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/190 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114768 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78100 "Passed tests") | | [❌ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7b9762d8-1837-454f-a5d1-fe85ae051b7f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/181 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32875 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78791 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108557 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33361 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137971 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/251 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/240 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106058 "Found 1 new test failure: imported/blink/fast/pagination/viewport-y-horizontal-bt-rtl.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/282 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111111 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105797 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27015 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/190 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29665 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/52435 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/298 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/61815 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/207 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/280 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/257 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -176,6 +214,10 @@ void setOpaque(bool opaque)
+                if (RefPtr client = protectedThis->m_client.get()) {
+                    auto [center, extents] = computeCenterAndExtents(updateRequest.parts, updateRequest.instanceTransforms);
+                    client->didUpdateBoundingBox(protectedThis.get(), center, extents);
+                }
```
This might be more correct. But I want to make sure we're mindful about breaking the parallel between this and the ModelProcess implementation (and potentially update the ModelProcess one when it makes sense). Especially if down the line this can help us avoid a third computeSRT implementation :) 

In ModelProcess land:
* The RKEntity computes the center and extents
* The Player(Proxy) keeps track of the original center+extents (to compute transforms later)
* `client->didUpdateBoundingBox(...)` only gets called on load right after `didFinishLoading()`

Could we do the computation in the ModelAdded callback based on the `addRequest`, and maybe move the code to DDMesh?

mwyrzykowski replied:
I think `didFinishLoading` in DDModelPlayer.mm:
```
                if (RefPtr client = protectedThis->m_client.get())
                    client->didFinishLoading(protectedThis.get());
```
is called too early.

At the point `addMesh` returns, we aren't aware of the model's geometry as addMesh only contains the layout of the geometry and not the geometry itself.

Would the concern be addressed if we moved `didFinishLoading` to the updateMesh path, only calling it upon the first update?

mwyrzykowski replied:
I can certainly move the AABB calculation to DDMesh as well

mwyrzykowski replied:
Fixed the `didFinishLoading` issue in https://github.com/WebKit/WebKit/pull/52590

mwyrzykowski replied:
@etiennesegonzac I adopted the visionOS style and only send `didUpdateBoundingBox` right after `didFinishLoading`

Also moved the computations to the DDMeshProxy

mwyrzykowski commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -141,75 +141,78 @@ void setOpaque(bool opaque)
 {
     RefPtr corePage = m_page.get();
     m_modelLoader = nil;
-    if (!m_modelLoader) {
```
this is mostly removed indentation because the conditional was silly, `m_modelLoader` is set to nil on the line above

mwyrzykowski commented on Source/WebCore/Modules/model-element/DDModelPlayer.mm:
```diff
@@ -141,75 +141,78 @@ void setOpaque(bool opaque)
+                protectedThis->m_didFinishLoading = true;
+                client->didFinishLoading(protectedThis.get());
+                auto [simdCenter, simdExtents] = protectedThis->m_currentModel->getCenterAndExtents();
+                client->didUpdateBoundingBox(protectedThis.get(), FloatPoint3D(simdCenter.x, simdCenter.y, simdCenter.z), FloatPoint3D(simdExtents.x, simdExtents.y, simdExtents.z));
```
matched the visionOS implementation until we decide its inaccurate

mwyrzykowski committed:
[macOS + iOS Model] Compute bounding box when geometry is updated
https://bugs.webkit.org/show_bug.cgi?id=300410
rdar://162233357

Reviewed by Etienne Segonzac.

Compute bounding box when the mesh geometry is updated.

Test: model-element/model-element-bounding-box.html passes locally after this change

* Source/WebCore/Modules/model-element/DDModelPlayer.mm:
(WebCore::computeCenterAndExtents):
(WebCore::DDModelPlayer::load):

Canonical link: https://commits.webkit.org/302194@main
